pennylane_lightning/core/jaxpr_adjoint.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""This module contains utility functions for working with JAX Jaxprs in the context of adjoint differentiation."""
+
+from typing import Sequence, Tuple
+
+import jax
+import numpy as np
+import pennylane as qml
+from pennylane.capture.primitives import AbstractMeasurement
+from pennylane.typing import TensorLike
+
+
+def validate_args_tangents(
+    args: Sequence[TensorLike], tangents: Sequence[TensorLike]
+) -> Tuple[TensorLike]:
+    """Ensure args and tangents provided to the JAXPR are consistent.
+
+    Args:
+        args (Sequence[TensorLike]): the arguments to the JAXPR
+        tangents (Sequence[TensorLike]): the tangents to the JAXPR
+
+    Returns:
+        Tuple[TensorLike]: the validated tangents
+    """
+
+    if len(args) != len(tangents):
+        raise ValueError("The number of arguments and tangents must match")
+
+    def _make_zero(tan, arg):
+        return jax.lax.zeros_like_array(arg) if isinstance(tan, jax.interpreters.ad.Zero) else tan
+
+    tangents = tuple(map(_make_zero, tangents, args))
+
+    for tan in tangents:
+        if jax.numpy.issubdtype(jax.numpy.asarray(tan).dtype, jax.numpy.integer):
+            raise ValueError("Tangents cannot be of integer type")
```

Comment:
```suggestion
            # will be able to support ad.Zero integer tangents when can track trainable parameters
            raise ValueError("Tangents cannot be of integer type yet.")
```
In the future, we should be fine to have `ad.Zero` integer tangents. But that will take being able to track trainable parameters. So just adding in a `"yet"` to indicate we can get around to it eventually and is not a fundamental limitation.